Governance initiated transfers (#698)
vegaprotocol/documentation: docs/tutorials/proposals/asset-transfer-proposal.md

The **terms** of the proposal describe the details of the proposal, including closing and enactment timestamps, and details about the proposed transfer, including what accounts the transfer is to move assets between, and optionally information about rewards to fund with the transfer, if that's the intent.
* **Closing timestamp**, the date and time when voting closes, must be within a range of <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.minClose" hideName={true}/> and <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.maxClose" hideName={true}/>, from when it was proposed.
* **Enactment timestamp**, the date and time when the transfer is to happen, must be within a range of <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.minEnact" hideName={true}/> and <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.maxClose" hideName={true}/>.

> **Amount** is how much to transfer, as whole number with the asset decimal places implied. For example, if the asset has a decimal place of 2 and the transfer is for 100, then the amount needs to be set at 10000. The maximum you can propose to transfer is governance.proposal.transfer.maxAmount. Before proposing, make sure the account you're transferring from exists and has a balance. The full amount may not be transferred if there isn't enough to transfer.

Reviewer: it's not strictly how much to transfer, it's a cap on how much to transfer. Without the formulas all of it is hard to understand.

Author: I'll link to the formulas - but I think they don't make it much easier to understand or at least not very accessible

**Fraction of balance** is used to determine what fraction of the source account's balance could be transferred, the biggest fraction that you can choose is <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.maxFraction" hideName={true}/>.

> **Governance transfer type** lets you determine if you want to try to transfer all or nothing, or to transfer as much as is possible up to the full amount.

Reviewer: that's a bit vague. I don't know if it can be understood without the formulas.

> ### Fields used to fund trading rewards

Reviewer: I think before we go into this, it should be mentioned that a transfer can be one off or recurring.

If you're proposing a transfer to fund rewards, there are extra fields to define.

> The `toAccountType` must be the account type that matches the reward category. For example, to propose that the 'average position' reward will pay out, you'll need to set the "to account type" type as `ACCOUNT_TYPE_REWARD_AVERAGE_POSITION`, and then choose the complementary reward category, known as the `metric`. The asset you choose then determines which market(s) the reward targets.

Reviewer: maybe I've missed it in here but is the whole what goes into source/destination explained here somewhere?

> The `toAccountType` must be the account type that matches the reward category. For example, to propose that the 'average position' reward will pay out, you'll need to set the "to account type" type as `ACCOUNT_TYPE_REWARD_AVERAGE_POSITION`, and then choose the complementary reward category, known as the `metric`. The asset you choose then determines which market(s) the reward targets.

Author: ```suggestion
The `destinationAccountType` must be the account type that matches the reward category. For example, to propose that the 'average position' reward will pay out, you'll need to set the "destination account type" as `ACCOUNT_TYPE_REWARD_AVERAGE_POSITION`, and then choose the complementary reward category, known as the `metric`. The asset you choose then determines which market(s) the reward targets.
```

> The `to` field is used for public keys. In the case of network-managed accounts, the input they need is `0000000000000000000000000000000000000000000000000000000000000000`.

Reviewer: there's no `to` field, there's `destination` field and the above is also inaccurate. The source [destination] is the public key of the party when the source [destination] account type is a party account. Otherwise, it depends on the context - if we're transferring to/from market's insurance pool, then the source [destination] is the market ID.

> You will need to define the dispatch strategy, which includes the  metric, the length of time to measure performance, the asset used to evaluate performance, and other fields. 

Reviewer: only supported for recurring governance transfers.

| Dispatch strategy field | Description | Accepted values |
| ----------- | ----------- | ----------- |

> | `assetForMetric` | Asset that's used to evaluate how someone performs, such as the settlement asset for the market(s) relevant to the reward | Any asset enabled on Vega |

Reviewer: not sure what you mean by "such as the settlement asset", it's literally the settlement asset for the markets to be included, it can't be something else.

| `metric` | Specific reward category the transfer is funding | DISPATCH_METRIC_MAKER_FEES_PAID; DISPATCH_METRIC_MAKER_FEES_RECEIVED; DISPATCH_METRIC_LP_FEES_RECEIVED; DISPATCH_METRIC_MARKET_VALUE; DISPATCH_METRIC_AVERAGE_POSITION; DISPATCH_METRIC_RELATIVE_RETURN; DISPATCH_METRIC_RETURN_VOLATILITY; DISPATCH_METRIC_VALIDATOR_RANKING |

> | `markets` | Optional: Used to choose which market(s) are in scope | Any trading market's ID |

Reviewer: if left empty it implies all the markets that are settled in the asset for metric.

| `staking_requirement` | Optional: Sets a minimum number of VEGA tokens that need to be staked for a party to be considered eligible for the reward | Number, if omitted it defaults to 0 |
| `notional_time_weighted_average_position_requirement` | Optional: Sets a minimum notional TWAP required for a party to be considered eligible to receive rewards | Defaults to 0 | 
| `windowLength` | Number of epochs in which performance against the reward metric is measured | Any number between 1 and 100 |
| `lock_period` | Number of epochs to keep earned rewards in the recipient's reward vesting account before moving to their vested account |

> | `entityScope` | defines the entities within scope | Currently ENTITY_SCOPE_INDIVIDUALS is the only option |

Reviewer: is this the final decision?

      "closingTimestamp": 1697283340,
      "enactmentTimestamp": 1697283400,
      "newTransfer": {
        "changes": {
          "fromAccountType": "ACCOUNT_TYPE_NETWORK_TREASURY",
          "transferType": "GOVERNANCE_TRANSFER_TYPE_BEST_EFFORT",
          "amount": "10000000",
          "asset": "b340c130096819428a62e5df407fd6abe66e444b89ad64f670beb98621c9c663",
          "fractionOfBalance": "0.1",
          "toAccountType": "ACCOUNT_TYPE_REWARD_AVERAGE_POSITION",
          "to": "0000000000000000000000000000000000000000000000000000000000000000",
          "recurring": {
            "startEpoch": 116515,
            "endEpoch": 116615
            }
          }
        }
      }
    }
}
```
  </TabItem>
  <TabItem value="cmd" label="Command line (Linux / OSX)">

  <TerminalInstructions />

```
./vegawallet transaction send --wallet YOUR_WALLETNAME --pubkey YOUR_PUBLIC_KEY --network NETWORK_NAME '{
 "proposalSubmission": {
  "rationale": {
   "title": "Propose moving assets",
   "description": "Here is a description about why I want to transfer"
  },
  "terms": {
    "closingTimestamp": 10000000,
    "enactmentTimestamp": 10000100,
    "newTransfer": {
    "changes": {

>      "fromAccountType": "ACCOUNT_TYPE_NETWORK_TREASURY",

Author: ```suggestion
     "sourceType": "ACCOUNT_TYPE_NETWORK_TREASURY",
```

     "transferType": "GOVERNANCE_TRANSFER_TYPE_BEST_EFFORT",
     "amount": "10000000",
          "asset": "ASSET_ID",
          "fractionOfBalance": "0.1",
          "toAccountType": "ACCOUNT_TYPE_REWARD_AVERAGE_POSITION",
          "to": "0000000000000000000000000000000000000000000000000000000000000000",
          "recurring": {
            "startEpoch": 1111110,
            "endEpoch": 1111111,
     }
    }
   },
  }
 }
}'
``` 
</TabItem>
  <TabItem value="win" label="Command line (Windows)">
    <TerminalInstructions />   

```
vegawallet.exe transaction send --wallet YOUR_WALLETNAME --pubkey YOUR_PUBLIC_KEY --network NETWORK_NAME ^
"{^
\"proposalSubmission\": {^
 \"rationale\": {^
  \"title\": \"Propose moving assets\",^
  \"description\": \"Here is a description about why I want to transfer\"^
 },^
 \"terms\": {^
  \"closingTimestamp\": \"10000000\",^
  \"enactmentTimestamp\": \"10000100\",^
  \"newTransfer\": {^
   \"changes\": {^

>     \"fromAccountType\": \"ACCOUNT_TYPE_NETWORK_TREASURY\",^

Author: ```suggestion
    \"sourceType\": \"ACCOUNT_TYPE_NETWORK_TREASURY\",^
```

The **terms** of the proposal describe the details of the proposal, including closing and enactment timestamps, and details about the proposed transfer, including what accounts the transfer is to move assets between, and optionally information about rewards to fund with the transfer, if that's the intent.
* **Closing timestamp**, the date and time when voting closes, must be within a range of <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.minClose" hideName={true}/> and <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.maxClose" hideName={true}/>, from when it was proposed.
* **Enactment timestamp**, the date and time when the transfer is to happen, must be within a range of <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.minEnact" hideName={true}/> and <NetworkParameter frontMatter={frontMatter} param="governance.proposal.transfer.maxClose" hideName={true}/>.

> For **one-off transfers** use `deliverOn` to set a delivery date/time for when the transfer arrives with the recipient account. `deliverOn` only accepts Unix time in nanoseconds. Setting it to 0 means the transfer will be completed immediately. Note: when you delay a transfer, the amount leaves the origin account immediately but is not delivered until the date/time you chose. A one-off transfer **cannot be cancelled**, regardless of when the transfer is scheduled to arrive.

Reviewer: note that this only happens after the proposal is enacted.

For **recurring transfers**, such as for funding rewards, you'll need to include the following information: 
* `startEpoch`: The number of the epoch in which you want the first transfer to be made. It will initiate at the end of that epoch.
* `endEpoch`: The transfer will repeated indefinitely, unless you add this optional parameter to end the recurring transfer in a specified epoch.

> * `factor`: Written as a decimal less than 1.0. Factor is used to determine what portion of the full `amount` is transferred in each epoch. Think of it like a percentage, so the number you include, when multiplied by 100, will equal what percentage of the amount will be transferred each time.

Reviewer: there's no such thing for governance transfers.

| ----------- | ----------- | ----------- |
| `assetForMetric` | Asset that's used to evaluate how someone performs. Use the settlement asset for the market(s) relevant to the reward. Not required for market creation and validator ranking rewards | Any asset enabled on Vega |
| `metric` | Specific reward category the transfer is funding | DISPATCH_METRIC_MAKER_FEES_PAID; DISPATCH_METRIC_MAKER_FEES_RECEIVED; DISPATCH_METRIC_LP_FEES_RECEIVED; DISPATCH_METRIC_MARKET_VALUE; DISPATCH_METRIC_AVERAGE_POSITION; DISPATCH_METRIC_RELATIVE_RETURN; DISPATCH_METRIC_RETURN_VOLATILITY; DISPATCH_METRIC_VALIDATOR_RANKING |
| `markets` | Optional: Used to choose which market(s) are in scope. If left blank, all markets that are settled in the asset are included | Any trading market's ID |
| `staking_requirement` | Optional: Sets a minimum number of VEGA tokens that need to be staked for a party to be considered eligible for the reward | Number, if omitted it defaults to 0 |
| `notional_time_weighted_average_position_requirement` | Optional: Sets a minimum notional TWAP, measured for the asset metric, that's required for a party to be considered eligible to receive rewards | Defaults to 0 | 

> | `windowLength` | Number of epochs in which performance against the reward metric is measured | Any number between 1 and 100 |

Reviewer: why are some params camelCase and some snake_case?
